app/src/behavior.c

```diff
@@ -56,29 +58,58 @@ const struct device *z_impl_behavior_get_binding(const char *name) {
+int zmk_behavior_invoke_binding(const struct zmk_behavior_binding *src_binding,
+                                struct zmk_behavior_binding_event event, bool pressed) {
+    LOG_DBG("`zmk_behavior_invoke_binding` is deprecated. Please raise a "
+            "`zmk_behavior_binding_event` instead.");
+    return raise_zmk_behavior_binding_event((struct zmk_behavior_binding_event){
+        .binding = src_binding,
+        .layer = event.layer,
+        .position = event.position,
+        .timestamp = event.timestamp,
+        .type = pressed ? PRESS : RELEASE,
+#if IS_ENABLED(CONFIG_ZMK_SPLIT)
+        .source = event.source,
+#endif
+    });
+}
+
+// TODO: Pass the event in as pointers, rather than copying in the struct
+// Make this change as part of a breaking release
 static int invoke_locally(struct zmk_behavior_binding *binding,
-                          struct zmk_behavior_binding_event event, bool pressed) {
-    if (pressed) {
-        return behavior_keymap_binding_pressed(binding, event);
-    } else {
-        return behavior_keymap_binding_released(binding, event);
+                          struct zmk_behavior_binding_event *event) {
+    switch (event->type) {
+    case PRESS:
+        return behavior_keymap_binding_pressed(binding, *event);
+    case RELEASE:
+        return behavior_keymap_binding_released(binding, *event);
+    case SENSOR:
+        return behavior_sensor_keymap_binding_process(binding, *event);
+    default:
+        return -EINVAL;
     }
 }
 
-int zmk_behavior_invoke_binding(const struct zmk_behavior_binding *src_binding,
-                                struct zmk_behavior_binding_event event, bool pressed) {
-    // We want to make a copy of this, since it may be converted from
-    // relative to absolute before being invoked
-    struct zmk_behavior_binding binding = *src_binding;
-
-    const struct device *behavior = zmk_behavior_get_binding(binding.behavior_dev);
+int behavior_listener(const zmk_event_t *eh) {
+    struct zmk_behavior_binding_event *event = as_zmk_behavior_binding_event(eh);
+    if (event == NULL) {
+        return -EINVAL;
+    }
 
+    const struct device *behavior = zmk_behavior_get_binding(event->binding->behavior_dev);
     if (!behavior) {
-        LOG_WRN("No behavior assigned to %d on layer %d", event.position, event.layer);
-        return 1;
+        LOG_WRN("No behavior assigned to %d on layer %d", event->position, event->layer);
+        return 0;
     }
 
-    int err = behavior_keymap_binding_convert_central_state_dependent_params(&binding, event);
+    /*
+     * Copying the behavior is necessary so that
+     * behavior_keymap_binding_convert_central_state_dependent_params (used for e.g. certain
+     * RGB behaviors) does not modify it, e.g. if the behavior is stored in a combo_cfg
+     */
+    struct zmk_behavior_binding binding = *(event->binding);
+    int err = behavior_keymap_binding_convert_central_state_dependent_params(&binding, *event);
```

Comment:
Is this still making a copy somewhere? I remember I had to tweak the signature of `zmk_behavior_invoke_binding` so that it used a struct that was copy-by-value, rather than a pointer

Reply:
You are right, I suppose L111 should be making a copy, similar to what was happening in L72 before. It'd be good to confirm via testing to be sure -- I was testing backlight behavior with nano LEDs on a simple split.

Reply:
Yup, that dereference like that will copy. No concerns there.